test: refactor recursive test cases

meskill opened the pull request:
**Summary:**  

I've noticed that we have multiple test for recursion but some of them were testing basically the same logic and some were not properly testing it, input types were not covered. In this pr I replaced test cases for recursion checks with new set of tests that covers both output/input types and proper assert for validation error. 

Also did small refactoring in the code to remove meaningless parts that are already covered by some other checks

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

tusharmath reviewed (changes requested):
Seem like we are fixing something and not just adding tests. Also we have renamed some of the existing tests making it harder to identify what changed. Can you revert that and also describe overall what did we fix here?

meskill commented:
> Seem like we are fixing something and not just adding tests. Also we have renamed some of the existing tests making it harder to identify what changed. Can you revert that and also describe overall what did we fix here?

I've updated the pr description to explain the changes
